src/BleOtaUploader.cpp

```diff
@@ -1,288 +1,499 @@
         return;
     }
-    memcpy(&firmwareLength, data, length);
+   
+    const bool withCompression = req.flags.compression;
+    if (withCompression)
+    {
+        if (not _decompressor.isSupported())
+        {
+            handleError(BleOtaStatus::CompressionNotSupported);
+            return;
+        }
+        _decompressor.begin(req.compressedSize);
+    }
+    _decompressor.setEnable(withCompression);
 
-    if (storage == nullptr or not storage->open(firmwareLength))
+    const bool withChecksum = req.flags.checksum;
+    if (withChecksum)
     {
-        firmwareLength = 0;
-        handleError(INTERNAL_STORAGE_ERROR);
-        return;
+        if (not _checksum.isSupported())
+        {
+            handleError(BleOtaStatus::ChecksumNotSupported);
+            return;
+        }
+        _checksum.begin();
     }
+    _checksum.setEnable(withChecksum);
 
-    if (storage->maxSize() and firmwareLength > storage->maxSize())
+    if (_signature.isEnabled())
     {
-        terminateUpload();
-        handleError(INCORRECT_FIRMWARE_SIZE);
+        _signature.begin();
+    }
+
+    const auto status = _storage.open(req.firmwareSize);
+    if (status != BleOtaStatus::Ok)
+    {
+        handleError(status);
         return;
     }
 
-    uploading = true;
-    crc.restart();
+    _state = State::Upload;
 
-    #ifndef BLE_OTA_NO_BUFFER
-    buffer.clear();
-    uint32_t bufferSize = withBuffer ? BLE_OTA_BUFFER_SIZE : 0;
-    #else
-    uint32_t bufferSize = 0;
-    #endif
-    BeginResponse resp{OK, BLE_OTA_ATTRIBUTE_SIZE, bufferSize};
-    send((const uint8_t*)(&resp), sizeof(BeginResponse));
+    const uint32_t packageSize = min(req.packageSize, BLE_OTA_PACKAGE_SIZE);
+    const uint32_t bufferSize = _buffer.begin(req.bufferSize, packageSize);
+    BLE_OTA_LOG(TAG, "Send BeginResp: size: (pkg: %lu, buff: %lu)", packageSize, bufferSize);
+    sendMessage(BleOtaBeginResp{packageSize, bufferSize});
 
-    ArduinoBleOTA.uploadCallbacks->onBegin(firmwareLength);
+    _uploadCallbacks->handleUploadBegin();
 }
 
-void BleOtaUploader::handlePackage(const uint8_t* data, size_t length)
+void BleOtaUploader::handlePackageReq(const BleOtaPackageReq& req)
 {
-    if (not uploading)
+    BLE_OTA_LOG(TAG, "Handle PackageReq");
+
+    if (_state != State::Upload)
+    {
+        handleError(_state == State::Terminate ? _terminateCode : BleOtaStatus::UploadStopped);
         return;
+    }
 
-    #ifndef BLE_OTA_NO_BUFFER
-    const bool sendResponse = not withBuffer or buffer.size() + length > BLE_OTA_BUFFER_SIZE;
-    #else
-    const bool sendResponse = true;
-    #endif
+#ifndef BLE_OTA_NO_BUFFER
+    BleOtaStatus status;
+    if (_buffer.isEnabled())
+    {
+        status = flushBuffer();
+        if (status == BleOtaStatus::Ok)
+            status = _buffer.push(req.data, req.size);
+    }
+    else
+    {
+        status = pushFirmware(req.data, req.size);
+    }
+#else
+    const auto status = pushFirmware(req.data, req.size);
+#endif
```

Comment:
```suggestion
#ifdef BLE_OTA_NO_BUFFER
    const auto status = pushFirmware(req.data, req.size);
#else
    BleOtaStatus status;
    if (_buffer.isEnabled())
    {
        status = flushBuffer();
        if (status == BleOtaStatus::Ok)
            status = _buffer.push(req.data, req.size);
    }
    else
    {
        status = pushFirmware(req.data, req.size);
    }
#endif
```

Reply:
I'm using `#ifndef BLE_OTA_NO_BUFFER` everywhere in this file. Maybe leave it as is in order to have some logic consistency?